### PR TITLE
sitemap (prototype)

### PR DESCRIPTION
**Pull request by raymondng**
A prototype/silver-bullet for quick feedback/learning. Does not scale and may not work.

The plugin currently is no-op, but the admin endpoint should be functional, though it functions in a way that (again) does not scale. You should consider it a temporary endpoint.

@andrewconner please take a look. don't merge yet. tested locally -- will need to add test, etc.

### CODE REVIEW

**Comment by andrewconner** [comment]
Wahoo! + @eishay

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+class SiteMapGenerator @Inject() (
+    airbrake: AirbrakeNotifier,
+    db: Database,
+    userRepo: UserRepo,
+    libraryRepo: LibraryRepo,
+    libraryCommander: LibraryCommander) extends Logging {
+
+  // prototype/silver-bullet implementation -- will learn and improve
+  def generate(): Future[Elem] = Future.successful {
+    val libraries = db.readOnlyReplica { implicit ro => libraryRepo.getAllPublishedLibraries().take(50000) } // ahem does not scale
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala:70]: paginate?

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    // may want to add warning when > 40K
+
+    // batch, optimize
+    val ownerIds = CollectionHelpers.dedupBy(libraries.map(_.ownerId))(id => id)
+    val owners = db.readOnlyMaster { implicit ro => userRepo.getUsers(ownerIds) } // cached
+    val paths = libraries.filter { lib => owners.get(lib.ownerId).isDefined } map { lib =>
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala]: we should not put libraries with three or less keeps in them. lets filter them out by count

**Comment by eishay** [comment]
i assume we don't have an endpoint yet and there's no ping to google, right?

**Comment by raymondng** [comment]
Right. Nothing in prod right now. No worries.

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala]: Agreed.

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      val owner = owners(lib.ownerId)
+      s"https://kifi.com${Library.formatLibraryPath(owner.username, owner.externalId, lib.slug)}"
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala]: `https://www.kifi.com`

**Comment by eishay** [comment]
also, it seems like the sitemap should be persisted (s3?) and cached for very fast retrieval.

**Comment by eishay** [comment]
Note that "The limit is 50,000 URLs or a maximum size of 10MB uncompressed".
http://googlewebmastercentral.blogspot.com/2014/10/best-practices-for-xml-sitemaps-rssatom.html
so we need to check for size as we grow it or put a safe distance from it though not make it too small "A common mistake is to put only a handful of URLs into each XML sitemap file, which usually makes it harder for Google to download all of these XML sitemaps in a reasonable time".

**Comment by eishay** [comment]
also please look at the "XML Sitemap Index" http://www.sitemaps.org/protocol.html for the chunked sitemaps. I think we should start with one sitemap asap (for quick user growth, super important to do it quick) and rss for each library, then soon after have the chunked sitemap with the index so we can scale it.

**Comment by raymondng** [comment]
Thanks for the feedback! Yes that's the idea -- to do a quick prototype, and to (benefit &) learn from it.
I am aware of the 50K limit (and we currently have only 500 custom published libraries -- lots and lots of headroom), so no real need for chunking right now ... and you may have noticed comments like `may want to add warning when > 40K` and `generate and save to disk, maybe` on the list of things to do. In reality I think the code will be "ready" way before we hit 50K limit, but if we grow faster than that, I'd be happy to jump in with both feet and take care of things.

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala (hunk 3 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala]: :)

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala (hunk 2 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala]: what's the downside? 3 keeps can also be good/useful, right?

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala (hunk 1 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala:70]: yes we'll need it going forward

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala]: Rough proxy for quality. Less than 3 likely doesn't have useful content to extract anyways.

Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala]: Thanks for the clarifications. No worries.

For the record, I do have a slightly different perspective -- I don't know if we are able to judge (or in the position to judge) quality, and I don't know if we can say for sure that a library with 3 keeps is not useful. Nonetheless let's do this and we'll iterate.

**Comment by raymondng** [comment]
Thanks for feedback. Will merge and take it out for a spin.

**Pull request merged**